sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -188,6 +188,9 @@ class Dataset[T] private[sql](
     }
   }
 
+  // Wrap analyzed logical plan with an analysis barrier so we won't traverse/resolve it again.
+  @transient private val planWithBarrier: LogicalPlan = AnalysisBarrier(logicalPlan)
```

Comment:
shall we just add `AnalysisBarrier` when constructing `logicalPlan` instead of adding a new field `planWithBarrier`?